### PR TITLE
Add code to restrict API publish,delete and import container based on permission, #PG-3878

### PR DESCRIPTION
**Pull request by AltamashShaikh**
### Description:

Add code to restrict API publish,delete and import container based on permission
Fixes: #PG-3878

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Review by snake14 — APPROVED** [review]
I haven't had a chance to functional test, but the changes look good. I'm a little unsure about creating FakeAccessTagManager instead of testing against AccessValidator->checkUseCustomTemplatesCapability(), but I'm sure you had good reason.

File: tests/Integration/APITest.php [hunk 1]
```diff
@@ -1285,7 +1294,7 @@ protected function setAdminUser()
     public function provideContainerConfig()
     {
         return array(
-            'Piwik\Access' => new FakeAccess()
+            'Piwik\Access' => new FakeAccessTagManager()
```
Comment (AltamashShaikh) [inline tests/Integration/APITest.php:1297]: @snake14 We were already using a `FakeAccess` from core but that function didn't check if you have `tagmanager_use_custom_templates` capability or not instead returns null if you have write access, so I extended the core mock class to check for `tagmanager_use_custom_templates` capability if requested

**Comment by AltamashShaikh** [comment]
> I haven't had a chance to functional test, but the changes look good. I'm a little unsure about creating FakeAccessTagManager instead of testing against AccessValidator->checkUseCustomTemplatesCapability(), but I'm sure you had good reason.

To test this
1. Create a user with write permissions
2. Create an auth_token for that user
3. Try to execute this below curlRequest
```
curl -X POST "https://localhost.matomo.com//index.php?date=yesterday&module=API&format=json&idContainer={idContainer}&idContainerVersion={idContainerVersion}&environment=dev&method=TagManager.publishContainerVersion&segment=&idSite={idsite}&period=day&token_auth={auth_tiken}" --insecure
```

**Comment by snake14** [comment]
> > I haven't had a chance to functional test, but the changes look good. I'm a little unsure about creating FakeAccessTagManager instead of testing against AccessValidator->checkUseCustomTemplatesCapability(), but I'm sure you had good reason.
> 
> To test this
> 
> 1. Create a user with write permissions
> 2. Create an auth_token for that user
> 3. Try to execute this below curlRequest
> 
> ```
> curl -X POST "https://localhost.matomo.com//index.php?date=yesterday&module=API&format=json&idContainer={idContainer}&idContainerVersion={idContainerVersion}&environment=dev&method=TagManager.publishContainerVersion&segment=&idSite={idsite}&period=day&token_auth={auth_tiken}" --insecure
> ```

Thanks @AltamashShaikh . I confirmed that I received the `You can't access this resource as it requires 'tagmanager_use_custom_templates' capability for the website id...` response for my write user.

**Pull request merged**